components/web_discovery/browser/payload_generator.cc

```diff
@@ -0,0 +1,203 @@
+        attribute_values_it->second.empty()) {
+      VLOG(1) << "Skipped clustered payload due to no values for root "
+                 "selector, action = "
+              << rule_group.action;
+      return std::nullopt;
+    }
+    if (rule.is_join) {
+      auto joined_payload = GenerateClusteredJoinedPayload(
+          kQueryActions.contains(rule_group.action), rule, matching_url_details,
+          attribute_values_it->second);
+      if (!joined_payload) {
+        VLOG(1) << "Skipped joined clustered payload, action = "
+                << rule_group.action;
+        return std::nullopt;
+      }
+      payload_rule_data = std::move(*joined_payload);
+    } else {
+      const auto* value = attribute_values_it->second[0].FindString(rule.key);
+      if (!value || value->empty()) {
+        VLOG(1) << "Skipped non-joined clustered payload, action = "
+                << rule_group.action;
+        return std::nullopt;
+      }
+      payload_rule_data = base::Value(*value);
+    }
+    inner_payload.Set(rule.key, std::move(payload_rule_data));
+  }
+  return CreatePayloadDict(rule_group, std::move(inner_payload));
+}
+
+void GenerateSinglePayloads(const ServerConfig& server_config,
+                            const PayloadRuleGroup& rule_group,
+                            const PageScrapeResult* scrape_result,
+                            std::vector<base::Value::Dict>& payloads) {
+  auto attribute_values_it = scrape_result->fields.find(rule_group.key);
+  if (attribute_values_it == scrape_result->fields.end()) {
+    return;
+  }
+  for (const auto& attribute_value : attribute_values_it->second) {
+    auto dict = attribute_value.Clone();
+    dict.Set(kCountryCodeFieldName, server_config.location);
+    payloads.push_back(CreatePayloadDict(rule_group, std::move(dict)));
+  }
+}
+
+}  // namespace
+
+std::vector<base::Value::Dict> GenerateQueryPayloads(
+    const ServerConfig& server_config,
+    const PatternsURLDetails* url_details,
+    std::unique_ptr<PageScrapeResult> scrape_result) {
+  std::vector<base::Value::Dict> payloads;
+  for (const auto& rule_group : url_details->payload_rule_groups) {
+    if (rule_group.rule_type == PayloadRuleType::kQuery &&
+        rule_group.result_type == PayloadResultType::kClustered) {
+      auto payload = GenerateClusteredPayload(rule_group, url_details,
+                                              scrape_result.get());
+      if (payload) {
+        payloads.push_back(std::move(*payload));
+      }
+    } else if (rule_group.rule_type == PayloadRuleType::kSingle &&
+               rule_group.result_type == PayloadResultType::kSingle) {
+      GenerateSinglePayloads(server_config, rule_group, scrape_result.get(),
+                             payloads);
+    }
+  }
+  return payloads;
+}
+
+base::Value::Dict GenerateAlivePayload(const ServerConfig& server_config,
+                                       std::string date_hour) {
+  base::Value::Dict payload;
```

Comment:
there is a slightly more readable way to do this, see https://source.chromium.org/chromium/chromium/src/+/main:base/values.h;l=421?q=values.h&ss=chromium